[SYCL][CUDA] Add async_group_copy free function implementation

FMarno opened the pull request:
Partial Implement for proposal at #4903.
This adds async_group_copy as a free function, implemented for `group` and `sub_group`.
This also adds optimizations for NVIDIA architectures of sm_80 and over, making it actually async.

Credit to @JackAKirk for most of the implementation.

Tests to follow shortly

JackAKirk commented:
Tests are ready to add to the llvm-test-suite here: https://github.com/intel/llvm-test-suite/compare/intel...JackAKirk:async-collective-tests?expand=1

alexbatashev commented on sycl/include/sycl/ext/oneapi/group_algorithm.hpp:
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+/// device_event which can be used to wait on the completion of the copy.
+/// Permitted types for dataT are all scalar and vector types, except boolean.
+template <typename Group, typename dataT>
+detail::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
+                    device_event>
```
This is C++14 feature, we can use it safely
```suggestion
std::enable_if_t<is_group_v<Group> && !detail::is_bool<dataT>::value,
                    device_event>
```

FMarno replied:
done

alexbatashev commented on sycl/include/sycl/ext/oneapi/group_algorithm.hpp:
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+                    device_event>
+async_group_copy(Group g, multi_ptr<T, SrcS> Src, multi_ptr<T, DestS> Dest,
+                 size_t NumElements, size_t Stride) {
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
```
I guess the idea here is that `bool` takes 1 byte? Wouldn't it be better then to replace `uint8_t` with `char`?

FMarno replied:
Yes, I think that is the idea. I've replaced that now.

alexbatashev commented on sycl/include/sycl/ext/oneapi/group_algorithm.hpp:
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
+  using VecT = detail::change_base_type_t<T, uint8_t>;
+  auto DestP = multi_ptr<VecT, DestS>(reinterpret_cast<VecT *>(Dest.get()));
```
Same as above. This `reinterpret_cast` is UB. Can we replace `uint8_t` with `char`?

FMarno replied:
done

alexbatashev commented on sycl/include/sycl/ext/oneapi/group_algorithm.hpp:
```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+  // https://github.com/intel/llvm/blob/sycl/libclc/generic/libspirv/async/wait_group_events.cl
+  // __spirv_ControlBarrier calls __syncthreads or __nvvm_bar_warp_sync
+  // https://github.com/intel/llvm/blob/sycl/libclc/ptx-nvidiacl/libspirv/synchronization/barrier.cl
+  (Events.ext_oneapi_wait(g), ...);
```
Folding expressions are C++ 17 feature. @romanovvlad should we guard those with macros for extensions?

romanovvlad replied:
It depends, there is a test: sycl/test/basic_tests/stdcpp_compat.cpp. If it passes there is no need to guard.

FMarno replied:
Works fine on my machine

alexbatashev reviewed (commented):
just a few nits

bader commented on libclc/generic/include/clc/async/common.h:
```diff
@@ -0,0 +1,33 @@
+    size = __spirv_SubgroupSize();                                             \
+    id = __spirv_SubgroupLocalInvocationId();                                  \
+  }                                                                            \
+  size_t i;                                                                    \
+                                                                               \
+  for (i = id; i < num_gentypes; i += size) {                                  \
+    dst[i * DST_STRIDE] = src[i * SRC_STRIDE];                                 \
+  }
```
```suggestion
                                                                               \
  for (size_t i = id; i < num_gentypes; i += size) {                           \
    dst[i * DST_STRIDE] = src[i * SRC_STRIDE];                                 \
  }
```

bader commented on libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:
```diff
@@ -0,0 +1,106 @@
+      const __attribute__((address_space(1))) TYPE *src, size_t num_gentypes,  \
+      size_t stride, event_t event) {                                          \
+    if (__nvvm_reflect("__CUDA_ARCH") >= 800) {                                \
+      size_t id, size;                                                         \
+      if (scope == Workgroup) {                                                \
+        id = (__spirv_WorkgroupSize_y() * __spirv_WorkgroupSize_x() *          \
+              __spirv_LocalInvocationId_z()) +                                 \
+             (__spirv_WorkgroupSize_x() * __spirv_LocalInvocationId_y()) +     \
+             __spirv_LocalInvocationId_x();                                    \
+        size = __spirv_WorkgroupSize_x() * __spirv_WorkgroupSize_y() *         \
+               __spirv_WorkgroupSize_z();                                      \
+      } else {                                                                 \
+        size = __spirv_SubgroupSize();                                         \
+        id = __spirv_SubgroupLocalInvocationId();                              \
+      }                                                                        \
```
This is exact copy of the code from `libclc/generic/include/clc/async/common.h`. I think it make sense to factor it out into a separate function/macro.

FMarno replied:
sure, updated.

bader commented on libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:
```diff
@@ -0,0 +1,106 @@
+        size = __spirv_SubgroupSize();                                         \
+        id = __spirv_SubgroupLocalInvocationId();                              \
+      }                                                                        \
+      size_t i;                                                                \
+      for (i = id; i < num_gentypes; i += size) {                              \
```
```suggestion
      for (size_t i = id; i < num_gentypes; i += size) {                       \
```

FMarno replied:
done

bader commented on libclc/ptx-nvidiacl/libspirv/async/async_work_group_strided_copy.cl:
```diff
@@ -0,0 +1,106 @@
+        size = __spirv_SubgroupSize();                                         \
+        id = __spirv_SubgroupLocalInvocationId();                              \
+      }                                                                        \
+      size_t i;                                                                \
+      for (i = id; i < num_gentypes; i += size) {                              \
```
```suggestion
      for (size_t i = id; i < num_gentypes; i += size) {                       \
```

FMarno replied:
done

bader commented on libclc/generic/include/clc/async/common.h:
```diff
@@ -0,0 +1,35 @@
+    SIZE = __spirv_SubgroupSize();                                             \
+    ID = __spirv_SubgroupLocalInvocationId();                                  \
+  }
+
+
```
```suggestion

```

bader commented:
/verify with https://github.com/intel/llvm-test-suite/pull/552

JackAKirk commented:
> /verify with [intel/llvm-test-suite#552](https://github.com/intel/llvm-test-suite/pull/552)

I've updated #552 reflecting the latest commits in this PR.

Pennycook commented:
I don't think we should merge this before the spec (https://github.com/intel/llvm/pull/4950) is merged, and I don't think we should mark it as a final/implemented/supported extension until we get clarification from Khronos on whether group algorithms act as synchronization points (https://gitlab.khronos.org/sycl/Specification/-/issues/576).

If we want to merge this soon, it may be worth considering putting it in the `experimental::` namespace.  That would make the feature available for testing, and we could take things out of `experimental::` once we've reached consensus on the expected synchronization behavior.  Tagging @gmlueck for comment, since he's also a reviewer on https://github.com/intel/llvm/pull/4950.

gmlueck commented:
> Moved async_group_copy to experimental

If the implementation is moved to "experimental", then the API spec in #4903 should also be updated to say the extension is experimental.

FMarno commented:
@gmlueck @Pennycook I've moved the implementation to the experimental namespace and updated the proposal in #4950. We'll update #4903 in the near future.

JackAKirk commented:
All Opencl types with 4, 8 or 16 byte alignment have been optimized in the cuda backend for sm_80.  The test is up to date here: https://github.com/intel/llvm-test-suite/pull/552

bader reviewed (changes requested):
@FMarno, could you resolve merge conflicts, please?

JackAKirk commented:
> @FMarno, could you resolve merge conflicts, please?

We've marked this as draft for the moment.  The `merge sycl` broke this impl due to some other changes from https://github.com/intel/llvm/commit/8a3bd1c5a9255fc4356f707009c9e5a463aca6ac.  I'll fix it now.
